[topic-gpio] drivers: gpio_stm32: update to use new GPIO API

erwango opened the pull request:
Update driver code to use new configuration flags such as GPIO_ACTIVE_LOW.
~~STM32F1 series not taken into account.~~
~~Only one board converted.~~
~~No backward compatibility: `gpio_configure` don't configure interrupts anymore~~

Based on:
- Provide arduino connector to nucleo_64 (#19628) and nucleo_144 (#19636)
- https://github.com/zephyrproject-rtos/zephyr/pull/19646 [topic-gpio] drivers/gpio: stm32: Rework configure function exit for dual core
- https://github.com/zephyrproject-rtos/zephyr/pull/19256 as new API assumes possibility to disabling and re-enabling interrupts


TODO:
- [x]  So far STM32F1 series are not taken into account. Extend to change to this series 
- [x] Get all STM32 based boards converted to the new API
 - ~~Done in dedicated PR https://github.com/zephyrproject-rtos/zephyr/pull/19607 in order to get useful Shippable result in current PR~~ Done in current PR
- [x] Test on all series
 - Identified nucleo_144 and nucleo_64 boards have different active states. ~~To be corrected in #19607~~ ~~Might be trickier than pure nucleo_64 vs nucleo_144. Need to check manuals in detail.~~ Done

EDIT: 
PR description rework after push on 09-09-19

mniestroj committed:
gpio: stm32: split helper functions from gpio_stm32_enable_int()

This patch doesn't change functionality, but is only related to improved
readability and reusability.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
gpio: stm32: statify gpio_stm32_enable_int()

This allows compiler to inline function body and reduce overall code
size.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
gpio: stm32: support disabling and reenabling interrupts on gpio pin

Up to now interrupts could be only configured once, with no way to
disable them in runtime.

Allow interrupts to be disabled in runtime and then properly reenabled
on user request. This allows to ignore interrupts when software is not
expecting them.

The improvement over previously reverted patch [1] is that we disable
interrupts only when we configure port for which interrupt line was
previously selected. This for example prevents to disable interrupts
line 2 in case PA2 was previously configured as interrupt source, but we
are currently configuring PB2 as output.

[1] 0951ce2d34b6 ("gpio: stm32: support disabling and reenabling
  interrupts on pin")

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

erwango commented:
@mniestroj, FYI

erwango commented:
Tested ok on `gpio_basic_api` but FATAL error when test is run w/o wire.
So this requires some more analysis.

erwango commented:
> Tested ok on gpio_basic_api but FATAL error when test is run w/o wire.
> So this requires some more analysis.

Seems to be intended behavior.

galak commented:
> > Tested ok on gpio_basic_api but FATAL error when test is run w/o wire.
> > So this requires some more analysis.
> 
> Seems to be intended behavior.

Yeah, seen the same thing myself

galak commented:
> * [ ]  Get all STM32 based boards converted to the new API

While maybe not the most efficient, here's an example to help convert the dts files over: 

`git grep -l \<st boards/arm | grep dts | xargs sed -i 's/GPIO_INT_ACTIVE_HIGH/GPIO_ACTIVE_LOW/g'`

galak commented:
Marking dnm since there are some todo items.

erwango committed:
boards: stm32: Update board definition using new GPIO api

Move GPIO_ACTIVE_INT_HIGH/LOW to GPIO_ACTIVE_HIGH/LOW.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
rechek

erwango commented:
recheck

erwango committed:
tests: gpio_basic_api: Enable test on stm32 boards

Add boards overlay to enable gpio_basic tests.
Selected boards should reflect various stm32 series.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

pabigot commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -41,24 +45,37 @@ static void gpio_stm32_isr(int line, void *arg)
+
 		*pincfg = STM32_PINCFG_MODE_OUTPUT;
+
+		if ((flags & GPIO_PUSH_PULL) != 0) {
```
This check is wrong: `GPIO_PUSH_PULL` has value zero, the branch will never be taken.

Review `include/dt-bindings/gpio/gpio.h`.

erwango replied:
Thanks!

pabigot commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -308,15 +325,16 @@ static int gpio_stm32_config(struct device *dev, int access_op,
-	if ((flags & GPIO_POL_MASK) == GPIO_POL_INV) {
-		/* hardware cannot invert signal */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
+		/* Simultaneous pin in/out mode is not supported */
```
Is disconnected (neither input nor output) supported?

erwango replied:
On some series (all except STM32F1), We can have alternate function or analog.
Analog is described as follows:
```
• The output buffer is disabled
• The Schmitt trigger input is deactivated, providing zero consumption for every analog
value of the I/O pin. The output of the Schmitt trigger is forced to a constant value (0).
• The weak pull-up and pull-down resistors are disabled
• Read access to the input data register gets the value “0”
```
So, I guess this matches disconnected.

pabigot replied:
Looking at the STM32L4x5/6 manual section 8.3.10 I agree with @mnkp: it appears that simultaneous in/out is supported whenever configured for output, and output-only (input disconnected) is not supported.

I think the helpful thing here would be to basically ignore `GPIO_INPUT`, so `GPIO_DISCONNECTED` and `GPIO_INPUT` are equivalent, and `GPIO_OUTPUT` and `GPIO_OUTPUT | GPIO_INPUT` are equivalent.

erwango replied:
@pabigot, @mnkp, so from this point of view, seems that this is the case indeed for all series. Thanks for the hint.
Though, input allows push-pull, which is not the case in analog. So I don't think we can have `GPIO_DISCONNECTED` and `GPIO_INPUT` put equivalent.

pabigot reviewed (changes requested):
Observed one flag test using the zero-valued variant, which needs to be fixed.  One or two other comments.

Checked on nucleo_l476rg, things passed.

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -369,6 +414,146 @@ static int gpio_stm32_read(struct device *dev, int access_op,
+	const struct gpio_stm32_config *cfg = dev->config->config_info;
+	GPIO_TypeDef *gpio = (GPIO_TypeDef *)cfg->base;
+
+	LL_GPIO_WriteOutputPort(gpio, mask & value);
```
This should be a read, modify, write operation.

erwango replied:
Right, thanks!

mnkp reviewed (changes requested):
I believe it would be better to have all stm32 changes in one PR instead of two. It makes it difficult to do the review since both PRs modify driver code. Also by not running checks on all the boards we may miss some issues. In case of Shippable timeout it's better to add a temporary patch from @galak that fixes the issue.

erwango commented:
@mnkp Thanks for review!

> It makes it difficult to do the review since both PRs modify driver code. 

This should not be the case PR #19067  only adds commit 717de74d7e2a91c3222f8d03979cfd93936bfebc which impact is limited to boards

> Also by not running checks on all the boards we may miss some issues. In case of Shippable timeout it's better to add a temporary patch from @galak that fixes the issue.

Agree it's better to actually get the results on all boards. @galak how can we make this happen?

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -369,6 +414,146 @@ static int gpio_stm32_read(struct device *dev, int access_op,
+	/*
+	 * On F1 series, using LL API requires a costly pin mask translation.
+	 * Skip it and use CMSIS API directly.
+	 * Extend its usage on other series as well.
```
Maybe the comment was not clear enough. 
Point is, due to STM32F1 specific GPIO IP, use of LL API requires a translation from u32 pin mask to mask of 'LL pins', which on STM32F1 are defined this way
```
#define LL_GPIO_PIN_10                   ((GPIO_BSRR_BS10 << GPIO_PIN_MASK_POS) | 0x04000004U)
```
Which, since BSRR/BRR are using a 'usual' pin mapping, would be again computed in LL function as:
```
WRITE_REG(GPIOx->BSRR, (PinMask >> GPIO_PIN_MASK_POS) & 0x0000FFFFU);
```
All these operations would be rather sub-optimal vs the direct use of following line:
```
WRITE_REG(gpio->BSRR, mask);
```

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -259,15 +325,16 @@ static int gpio_stm32_config(struct device *dev, int access_op,
 
-	if ((flags & GPIO_POL_MASK) == GPIO_POL_INV) {
-		/* hardware cannot invert signal */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
```
Configuration wise, there's a MODE register that make input and output mutually exclusive for most of series, by allowing following configurations:
```
Bits 2y:2y+1 MODERy[1:0]: Port x configuration bits (y = 0..15)
These bits are written by software to configure the I/O direction mode.
00: Input (reset state)
01: General purpose output mode
10: Alternate function mode
11: Analog mode
```
And STM32F1 (only exception I think):
```
MODEy[1:0]: Port x mode bits (y= 8 .. 15)
21:20, 17:16, 13:12,
These bits are written by software to configure the corresponding I/O port.
9:8, 5:4, 1:0
Refer to Table 20: Port bit configuration table.
00: Input mode (reset state)
01: Output mode, max speed 10 MHz.
10: Output mode, max speed 2 MHz.
11: Output mode, max speed 50 MHz.
```

This being said, I'm open to suggestions if I'm missing an alternate way for configuration.

pabigot commented:
@erwango As suggested to me by @carlescufi:
```
From e888dbf355ca8a63ec17ed725368ebb2ec79a200 Mon Sep 17 00:00:00 2001
From: Peter Bigot <peter.bigot@nordicsemi.no>
Date: Sun, 6 Oct 2019 15:05:28 -0500
Subject: [PATCH] DNM shippable.yml: increase builds to avoid timeout failures

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
---
 .shippable.yml | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/.shippable.yml b/.shippable.yml
index 605fd41be8..b50342a5ca 100644
--- a/.shippable.yml
+++ b/.shippable.yml
@@ -6,13 +6,20 @@ env:
     global:
         - ZEPHYR_SDK_INSTALL_DIR=/opt/sdk/zephyr-sdk-0.10.3
         - ZEPHYR_TOOLCHAIN_VARIANT=zephyr
-        - MATRIX_BUILDS="5"
+        - MATRIX_BUILDS="12"
     matrix:
         - MATRIX_BUILD="1"
         - MATRIX_BUILD="2"
         - MATRIX_BUILD="3"
         - MATRIX_BUILD="4"
         - MATRIX_BUILD="5"
+        - MATRIX_BUILD="6"
+        - MATRIX_BUILD="7"
+        - MATRIX_BUILD="8"
+        - MATRIX_BUILD="9"
+        - MATRIX_BUILD="10"
+        - MATRIX_BUILD="11"
+        - MATRIX_BUILD="12"
 
 build:
     cache: false
-- 
2.20.1
```

Also https://github.com/zephyrproject-rtos/zephyr/pull/19562#issuecomment-538075156

erwango commented:
@pabigot 

> @erwango As suggested to me by @carlescufi:

Thanks, will use that.

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -281,51 +348,29 @@ static int gpio_stm32_config(struct device *dev, int access_op,
-			return -ENOSYS;
+	if ((flags & GPIO_OUTPUT) != 0) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			gpio_stm32_write(dev, access_op, pin, 1U);
```
ok, thanks

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -188,7 +188,7 @@ int gpio_stm32_configure(u32_t *base_addr, int pin, int conf, int altf)
  * @brief Enable EXTI of the specific line
  */
-const int gpio_stm32_enable_int(int port, int pin)
+static int gpio_stm32_enable_int(int port, int pin)
```
Right, thanks.

erwango committed:
drivers/gpio: stm32: Update to new GPIO api

Update STM32 GPIO driver to support new GPIO API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/gpio: stm32: various clean up

Perform few clean up in stm32 gpio driver:
*Clean up uint32_t occurrences left over.
*Move function gpio_stm32_flags_to_conf from const to static
and remove useless parameter check
*Rework error handling in functoin gpio_stm32_config
*Remove gpio_stm32_int_enabled_port function and have direct call
to gpio_stm32_get_exti_source

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@mnkp, @pabigot, @ydamigos, @avisconti,

Thanks for your reviews.
Pushed a new version of the work. I think I've addressed most of the remarks related to the drivers.
Also I squashed 2 PR (driver and gpio) to get everything at one place and add a patch to help building all w/o timeout in shippable (so I will close https://github.com/zephyrproject-rtos/zephyr/pull/19607)
What I didn't do is to test, so I plan to do that tomorrow.

ydamigos commented on boards/arm/nucleo_g071rb/nucleo_g071rb.dts:
```diff
@@ -22,7 +23,7 @@
 		compatible = "gpio-leds";
 		green_led_1: led_4 {
-			gpios = <&gpioa 5 GPIO_INT_ACTIVE_LOW>;
+			gpios = <&gpioa 5 GPIO_ACTIVE_LOW>;
```
According to [STM32 Nucleo-64 boards with STM32G07xRB MCUs](https://www.st.com/content/ccc/resource/technical/document/user_manual/group1/5e/a5/b2/30/af/d4/45/d0/DM00452640/files/DM00452640.pdf/jcr:content/translations/en.DM00452640.pdf)
```
LD4 USER: this green LED is a user LED connected to Arduino signal D13 corresponding to STM32 I/O PA5. To light the LED a high-logic state ‘1’ has to be written in the corresponding GPIO.
```

mnkp commented:
> Pushed a new version of the work. I think I've addressed most of the remarks related to the drivers.

The new version doesn't seem to implement any changes to `gpio_stm32.c` driver requested during the review. Maybe it was not the latest version that got pushed?

erwango commented:
Indeed, odd, I'll have a look.

erwango committed:
boards: stm32: Fix buttons and leds configuration

On some stm32 based boards buttons and leds configuration
was wrong. Fix that.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

mnkp commented:
>     * `gpio_api_1pin`: Failed all tested boards so far. Investigating this is my next task

There is one issue I've noticed just recently that may cause 1 pin testcase to fail. The testcases testing the new port functions attempt to verify that the driver doesn't change other pins than the one that was requested. But the test doesn't take into account that the LED0 pin it's using for testing may be on the same port as UART used for the console. When reading port values the pin serving as UART RX will introduce inconsistency to the port value the testcase is expecting. If you see failures in any of `test_gpio_port_*` tests please ignore it for now. I'll provide a fix.

>     * `button`: Some boards button definitions are wrong.

The existing definitions in .dts files were unfortunately not always correct. It's best to fill in board DTS based on the board user manual / schematic.

erwango commented:
> > ```
> > * `gpio_api_1pin`: Failed all tested boards so far. Investigating this is my next task
> > ```
> 
> There is one issue I've noticed just recently that may cause 1 pin testcase to fail. The testcases testing the new port functions attempt to verify that the driver doesn't change other pins than the one that was requested. But the test doesn't take into account that the LED0 pin it's using for testing may be on the same port as UART used for the console. When reading port values the pin serving as UART RX will introduce inconsistency to the port value the testcase is expecting. If you see failures in any of `test_gpio_port_*` tests please ignore it for now. I'll provide a fix.
> 

Ok, thanks.

> > ```
> > * `button`: Some boards button definitions are wrong.
> > ```
> 
> The existing definitions in .dts files were unfortunately not always correct. It's best to fill in board DTS based on the board user manual / schematic.

Indeed, this wasn't clear, but this is in my todo list as well.

erwango commented:
>  * gpio_api_1pin: Failed all tested boards so far. Investigating this is my next task


Failed tests (when no conflict with other pins of the same port) are the following:
-  test_gpio_pin_configure_single_ended
- test_gpio_int_edge_falling
- test_gpio_int_edge_to_active
- test_gpio_int_edge_to_inactive

erwango committed:
drivers/gpio: stm32: exti: Clear triggers when not requested

Since it is now possible to disable/re-enable interrupts and
also to reconfigure an already configured interrupt, it is
now required to clear non requested triggers.
While it is not strictly requested, triggers are also cleared
when interrupt is disabled (assuming trigger should be configured
when interrupt is enabled).

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

fixup exti

erwango committed:
boards: stm32h747i_disco: Add arduino connector header

Add arduino connection dts description and update target
yaml files.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Pushed new version with fixes on tests failing in previous version .

Tested on following series: F0/F1/F3/F4/F7/G0/L0/L1/L4/WB

Remain to be tested: 
* F2: Need to find a board. Should be doable tomorrow.
* G4: Supported in `master` but not available on `topic-gpio` yet
* L1: Only board available is `96b_wistrio`, which I don't have. @Mani-Sadhasivam, can you perform tests on this board?


Test status:
* `gpio_basic_api`: Passed on all available boards except `stm32h747_disco_m7`, cf below
* `gpio_api_1pin`: Passed on available boards except `stm32h747_disco_m7`, cf below
* `button`: Passed on all available boards


Test failures on `stm32h747_disco_m7`:
* `gpio_basic_api`:
```
starting test - test_gpio_port
Validate device GPIOJ
Check GPIOJ output 9 connected to input 8
OUT 9 to IN 8 linkage works
- bits_physical
- pin_physical
- check_raw_output_levels
- check_input_levels
- bits_logical
- check_pulls

    Assertion failed at ../../src/test_gpio_port.c:389: check_pulls: (raw_in() not equal to false)
physical pull-down does not read low
FAIL - test_gpio_port
```
* `gpio_api_1pin`: 
```
starting test - test_gpio_pin_configure_push_pull
Running test on port=GPIOI, pin=12

    Assertion failed at ../../src/test_config.c:32: pin_get_raw_and_verify: (val_expected not equal to val_)
Test point 10: invalid pin get value
FAIL - test_gpio_pin_configure_push_pull
```

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -39,28 +43,55 @@ static void gpio_stm32_isr(int line, void *arg)
+	} else if  ((flags & GPIO_INPUT) != 0) {
+		/* Input */
+
+		if ((flags & GPIO_LINE_OPEN_DRAIN) != 0) {
```
We should implement this check directly in the API as an assert (guarding against a case when output is not enabled but `GPIO_LINE_OPEN_DRAIN` is set). Let's remove it from the driver.

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -211,47 +311,99 @@ const int gpio_stm32_enable_int(int port, int pin)
+/**
+ * @brief Get enabled GPIO port for EXTI of the specific pin number
+ */
+static int gpio_stm32_int_enabled_port(int pin)
```
The name of the function doesn't make it clear what the function does. It may be better to use `gpio_stm32_get_exti_source` directly (after moving there the check for an invalid pin).

erwango replied:
> (after moving there the check for an invalid pin).

Would be a relief to have this done in API.

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -211,47 +311,99 @@ const int gpio_stm32_enable_int(int port, int pin)
-	 * by writing value 0x5 instead of 0x7.
+	 * On F1 series, using LL API requires a costly pin mask translation.
+	 * Skip it and use CMSIS API directly.
+	 * Since valid on other series, extend on other series as well.
```
What about rewriting it as: ?
```
	 * Skip it and use CMSIS API directly. Valid also on other series.
```

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -211,47 +311,99 @@ const int gpio_stm32_enable_int(int port, int pin)
+	/*
+	 * On F1 series, using LL API requires a costly pin mask translation.
+	 * Skip it and use CMSIS API directly.
+	 * Extend its usage on other series as well.
```
```
	 * Skip it and use CMSIS API directly. Valid also on other series.
```
?

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -280,51 +426,19 @@ static int gpio_stm32_config(struct device *dev, int access_op,
+			gpio_stm32_port_set_bits_raw(dev, BIT(pin));
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			gpio_stm32_port_clear_bits_raw(dev, BIT(pin));
 		}
```
We should set the initial output value before we enable the output. Enabling the output is done by `gpio_stm32_configure` called on line 434.

mnkp commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -376,6 +490,75 @@ static int gpio_stm32_read(struct device *dev, int access_op,
+
+	if (stm32_exti_enable(pin) != 0) {
+		err = -EIO;
+		goto release_lock;
```
In a meantime we've called a few exti functions that seem to acquire/use resources: `stm32_exti_set_callback`, `gpio_stm32_enable_int`, `stm32_exti_trigger`. Do we need to do some clean up here?

erwango replied:
Actually `stm32_exti_enable` should be void.

erwango committed:
drivers/interrupt_controller: stm32: Clean set api from unused arg

Argument 'port' in stm32_exti_set_callback function is not
used, remove it.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/interrup_controller: stm32: stm32_exti_enable could be void

stm32_exti_enable was returning errors on line > 32 or line pointing
to non implemented line. Both conditions are hard-coded, hence there
is no use to detect them dynamically in the code.
Check them with assert. As a consequence, function could now be void.

Additionally, enable exti irq line only if both checks are passed.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32mp157c_dk2: Add led and button to dts

Add led and button to board description to enable basic gpio
tests.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Pushed new version taking into account last review from @nmkp and testing on F2 and MP1.

Tested on following series: F0/F1/F3/F4/F7/G0/H7/L0/L4/MP1/WB

Remain to be tested: 
* G4: Supported in `master` but not available on `topic-gpio` yet
* L1: Only board available is `96b_wistrio`, which I don't have. @Mani-Sadhasivam, can you perform tests on this board?


Test status:
* `gpio_basic_api`: Passed on all available boards except `stm32h747_disco_m7`, cf below
* `gpio_api_1pin`: Passed on available boards except `stm32h747_disco_m7`, and `stm32mp157c_dk2` cf below
* `button`: Passed on all available boards


Test failures on `stm32h747_disco_m7`:
* `gpio_basic_api`:
```
starting test - test_gpio_port
Validate device GPIOJ
Check GPIOJ output 9 connected to input 8
OUT 9 to IN 8 linkage works
- bits_physical
- pin_physical
- check_raw_output_levels
- check_input_levels
- bits_logical
- check_pulls

    Assertion failed at ../../src/test_gpio_port.c:389: check_pulls: (raw_in() not equal to false)
physical pull-down does not read low
FAIL - test_gpio_port
```

Playing with other boards, I've seen several times that this specific failure can happen when pin selected is in conflict with other usage. I've not been able to identify the conflict yet, but I heavily suspsect this issue is purely a configuration issue.

* `gpio_api_1pin`: 
```
starting test - test_gpio_pin_configure_push_pull
Running test on port=GPIOI, pin=12

    Assertion failed at ../../src/test_config.c:32: pin_get_raw_and_verify: (val_expected not equal to val_)
Test point 10: invalid pin get value
FAIL - test_gpio_pin_configure_push_pull
```
Can be PASSED if changing pin.

Test failures on `stm32mp157c_dk2`:
* `gpio_api_1pin`: Failing point relates to port testing. Due to use of Linux on Cortex-A, lot of pins are configured for various purposes (ethernet, display, ...), there is no empty port available for testing matching pins available on arduino connector. I'll have a new try once #19692 is fixed.

Mani-Sadhasivam reviewed (approved):
Tested the changes on 96Boards Wistrio which is based on STM32L1 SoC.

Mani-Sadhasivam commented:
@erwango Oops. I see couple of failures by executing the `gpio_basic_api` test. Earlier I just tested the `button` sample manually.

```
PASS - test_gpio_pin_read_write
===================================================================
starting test - test_gpio_callback_edge_high
callback triggered: 1
check enabled callback
PASS - test_gpio_callback_edge_high
===================================================================
starting test - test_gpio_callback_edge_low
FAIL - test_callback@68. config PIN_IN fail
    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_basic_api/src/test_callback_te

FAIL - test_gpio_callback_edge_low
===================================================================
starting test - test_gpio_callback_level_high
FAIL - test_callback@68. config PIN_IN fail
    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_basic_api/src/test_callback_te

FAIL - test_gpio_callback_level_high
===================================================================
starting test - test_gpio_callback_add_remove
callback_2 triggered: 1
callback_1 triggered: 1
callback_2 triggered: 1
PASS - test_gpio_callback_add_remove
===================================================================
starting test - test_gpio_callback_self_remove
callback_remove_self triggered: 1
callback_1 triggered: 1
callback_1 triggered: 1
PASS - test_gpio_callback_self_remove
===================================================================
starting test - test_gpio_callback_enable_disable
callback_2 triggered: 1
callback_1 triggered: 1
callback_2 triggered: 1
callback_1 triggered: 1
PASS - test_gpio_callback_enable_disable
===================================================================
starting test - test_gpio_callback_level_low
FAIL - test_callback@68. config PIN_IN fail
    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_basic_api/src/test_callback_te

FAIL - test_gpio_callback_level_low
===================================================================
Test suite gpio_basic_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

erwango commented:
@Mani-Sadhasivam, Thanks for this status. Can you check the port of the pins used in when testing `gpio_baisc_api` do not concflict port of pins used for other activity (like UART)?
Also, you can run new test `gpio_api_1pin`, which is also under `tests/drivers/gpio`

Mani-Sadhasivam commented:
@erwango 

> Can you check the port of the pins used in when testing gpio_baisc_api do not concflict port of pins used for other activity (like UART)?

Yes, these are dedicated GPIO pins. So there is no conflict.

> Also, you can run new test gpio_api_1pin, which is also under tests/drivers/gpio

For this test, only one testcase failed:

```
starting test - test_gpio_port_set_masked_get_active_low
Running test on port=GPIOA, pin=12
Step 1: Set logical, get logical and physical port value

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_port.c:41: )
Test point 0: invalid logical port get value
FAIL - test_gpio_port_set_masked_get_active_low
```

erwango commented:
> @erwango
> 
> > Can you check the port of the pins used in when testing gpio_baisc_api do not concflict port of pins used for other activity (like UART)?
> 
> Yes, these are dedicated GPIO pins. So there is no conflict.

Out of curiosity, which pins did you used?
Then, would you be able to analyze a bit the failure reason ?

> 
> > Also, you can run new test gpio_api_1pin, which is also under tests/drivers/gpio
> 
> For this test, only one testcase failed:
> 
> ```
> starting test - test_gpio_port_set_masked_get_active_low
> Running test on port=GPIOA, pin=12
> Step 1: Set logical, get logical and physical port value
> 
>     Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_port.c:41: )
> Test point 0: invalid logical port get value
> FAIL - test_gpio_port_set_masked_get_active_low
> ```

Tested pin is on GPIOA, which is same port as used by UART. 
Any way you can test on other port than GPIOA (even if there is no led attached), to see if status changed.

erwango commented:
@pabigot, @mnkp, shippable still times out despite the patch (even further increased up to 15 jobs). I think this is due to the fact I'm including commits that impact a lot of boards. But I also pushed these commits in master branch and they are now merged. 

Anyway we can rebase topic-gpio branch on recent master, so I can get rid of these commits in current PR and get more chances to get shippable completing the testing?

Mani-Sadhasivam commented:
@erwango I think something was messed up in my setup. Now, the `gpio_basic_test` succeeds but `gpio_api_1pin_test` still fails.

> Any way you can test on other port than GPIOA (even if there is no led attached), to see if status changed.

Tested PORTB pin 12, but same failure.

```
starting test - test_gpio_port_set_masked_get_active_low
Running test on port=GPIOB, pin=12
Step 1: Set logical, get logical and physical port value

    Assertion failed at ZEPHYR_BASE/tests/drivers/gpio/gpio_api_1pin/src/test_port.c:41: )
Test point 0: invalid logical port get value
FAIL - test_gpio_port_set_masked_get_active_low
```

Mani-Sadhasivam commented:
@erwango Looks like the failures are due to the pins owned by other peripherals as you said. So all good on the L1x side.

pabigot commented:
> @erwango Looks like the failures are due to the pins owned by other peripherals as you said. So all good on the L1x side.

Is that true for `gpio_basic_api` as well?  `gpio_api_1pin` is subject to false negatives due to other pins, but `gpio_basic_api` is not known to have any such problem: it only uses the two pins specified in the board overlay (which must be correctly pinmux-configured as GPIO).

pabigot commented:
> @pabigot, @mnkp, shippable still times out despite the patch (even further increased up to 15 jobs). I think this is due to the fact I'm including commits that impact a lot of boards. But I also pushed these commits in master branch and they are now merged.
> 
> Anyway we can rebase topic-gpio branch on recent master, so I can get rid of these commits in current PR and get more chances to get shippable completing the testing?

I don't have permission to push to this branch.  I have done a rebase (which required edits related to timeout parameters in the 2-pin test), available at https://github.com/pabigot/zephyr/commits/topic-gpio, if @carlescufi or somebody with maintainer permissions wishes to update.

```
tirzah[756]$ git push --force-with-lease=topic-gpio:r.upstream/topic-gpio 
Enumerating objects: 526, done.
Counting objects: 100% (526/526), done.
Delta compression using up to 32 threads
Compressing objects: 100% (303/303), done.
Writing objects: 100% (398/398), 74.00 KiB | 7.40 MiB/s, done.
Total 398 (delta 329), reused 114 (delta 90)
remote: Resolving deltas: 100% (329/329), completed with 115 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/topic-gpio.
remote: error: Cannot force-push to a protected branch
To github.com:/zephyrproject-rtos/zephyr
 ! [remote rejected]       topic-gpio -> topic-gpio (protected branch hook declined)
error: failed to push some refs to 'git@github.com:/zephyrproject-rtos/zephyr'
```

carlescufi commented:
@erwango 
The topic branch has been rebased. Could you please rebase this PR against it?

erwango commented:
@pabigot, @mnkp: shippable passed w/o boards modifications (cf https://github.com/zephyrproject-rtos/zephyr/pull/19764).

mnkp commented:
Is the `DNM` label still applicable?

erwango commented:
> Is the DNM label still applicable?

I still need to test on G4, but I should be able to do it in the afternoon, otherwise I think we can remove it.

erwango commented:
Status following last push:

Tested on all STM32 series.

### Test status:
* `gpio_basic_api`: Passed on all available boards except `stm32h747_disco_m7`, cf below
* `gpio_api_1pin`: Passed on available boards except `stm32h747_disco_m7`, and `stm32mp157c_dk2` cf below
* `button`: Passed on all available boards


#### Test failures on `stm32h747_disco_m7`:
* `gpio_basic_api`:
Playing with other boards, I've seen several times that this specific failure can happen when pin selected is in conflict with other usage. I've not been able to identify the conflict yet, but I heavily suspsect this issue is purely a configuration issue.
* `gpio_api_1pin`: 
Can be PASSED if changing pin.

STM32H7 port is still very basic for now (only GPIO, UART) and globally a wip.
Given this status and board complexity that might explain itself tests failures, I think it is safe to enable new GPIO API on this series.

#### Test failures on `stm32mp157c_dk2`:
* `gpio_api_1pin`: Failing point relates to port testing. Due to use of Linux on Cortex-A, lot of pins are configured for various purposes (ethernet, display, ...), there is no empty port available for testing matching pins available on arduino connector. I'll have a new try once #19692 is fixed.

Given board complexity board complexity that might explain tests failures, I think it is safe to enable new GPIO API on this series. 

#### Summary
Unless reviewers comment, I consider the work done on this PR.

erwango committed:
tests: gpio_api_1pin: Tune RAM console size for stm32mp157c_dk2


Test generates lengthy log.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
CODEWONERS: Update for gpio_stm32 driver

Change codeowner following driver indeep rework.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

carlescufi commented:
@erwango could you please rebase?

erwango commented:
> @erwango could you please rebase?

@carlescufi, done.

kklimek commented on drivers/interrupt_controller/exti_stm32.h:
```diff
@@ -44,10 +44,14 @@ void stm32_exti_disable(int line);
 	STM32_EXTI_TRIG_RISING  = 0x1,
 	/* trigger on falling endge */
 	STM32_EXTI_TRIG_FALLING = 0x2,
+	/* trigger on falling endge */
```
endge?
